src/main/java/core/basesyntax/model/io/DataConverterImpl.java

```
@@ -0,0 +1,69 @@
+package core.basesyntax.model.io;
+
+import core.basesyntax.model.FruitTransaction;
+import java.util.ArrayList;
+import java.util.List;
+
+public class DataConverterImpl implements DataConverter {
+
+    @Override
+    public List<FruitTransaction> convertToTransaction(List<String> lines) {
+        if (lines == null || lines.isEmpty()) {
+            throw new RuntimeException("Brak danych wejściowych do przetworzenia.");
+        }
+
+        List<FruitTransaction> transactions = new ArrayList<>();
+        int lineNumber = 0;
+
+        for (String line : lines) {
+            lineNumber++;
+
+            if (line == null || line.trim().isEmpty()) {
+                throw new RuntimeException("Wiersz " + lineNumber + " jest pusty.");
+            }
+
+            String[] parts = line.split(",");
+
+            if (parts.length != 3) {
+                throw new RuntimeException("Incorrect row format " + lineNumber
+                        + ": 3 fields expected, found " + parts.length);
+            }
+
+            String opCode = parts[0].trim();
+            String fruit = parts[1].trim();
+            String quantityStr = parts[2].trim();
+
+            if (opCode.isEmpty()) {
+                throw new RuntimeException("No operation code in the line " + lineNumber);
+            }
+
+            FruitTransaction.Operation op;
+            try {
+                op = FruitTransaction.Operation.fromCode(opCode);
+            } catch (IllegalArgumentException e) {
```

The unknown operation handling catches IllegalArgumentException and throws a RuntimeException but the message does not include the original CSV line and you don't pass the caught exception as the cause. The requirements say: "Validate and parse: Operation code must be one of b/s/p/r; error on invalid code." and also: "Do not ignore exceptions; rethrow as RuntimeException with informative messages and original cause." Include the original `line` in the message and pass `e` as the cause when rethrowing.